pkg/util/grpcclient/ratelimit.go

```diff
@@ -15,7 +15,10 @@ func NewRateLimiter(cfg *Config) grpc.UnaryClientInterceptor {
 	}
 	limiter := rate.NewLimiter(rate.Limit(cfg.RateLimit), burst)
 	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-		limiter.Wait(ctx)
+		err := limiter.Wait(ctx)
+		if err != nil {
+			return err
```

Comment:
When used in conjunction with `NewBackoffRetry()` (see `grpcclient.go`), the backoff expects a `ResourceExhausted` error code.

@bboreham what do you think returning that GRPC error here?

Reply:
That sounds like a bug.
"Error: I'm not going to wait because the context will expire before the wait finishes"
"Never mind, try again!"

Reply:
So @bboreham  / @pracucci  if i am understanding correctly what you are saying is that making this change is the wrong thing to do. Basically since the wait would not return a ResourceExhausted that would cause the BackoffRetry, if its being used, to now exit and no longer retry like we expect it to.

So if that is correct the right course of action is to just go back to what we had in originally here
https://github.com/cortexproject/cortex/pull/2008#discussion_r369464567
and ignore the error or should the BackoffRetry code check for another error code along with `ResourceExhausted`.

Reply:
I would not rollback but fix it. We may do:
`return status.New(codes.ResourceExhausted, err.Error())`

But I would be glad to also see a unit test on it.

Reply:
I'll look at adding a unit test for this and getting that change in place.

Reply:
Ok thanks! This looks like the very last thing left to me.

Reply:
@pracucci So i wrote up this unit test last night.....which might not be enough honestly. It's fine that it can validate that its a status with the expected result but an integration test would probably drive home the fact that it's necessary if you configure it with the backoffRetry. I just ran out of time.....I'll see if i can get to it over the weekend.  

```package grpcclient_test

import (
	"context"
	"testing"

	"github.com/cortexproject/cortex/pkg/util/grpcclient"
	assert2 "github.com/stretchr/testify/assert"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func TestRateLimiterFailureResultsInResourceExhaustedError(t *testing.T) {
	config := grpcclient.Config{
		RateLimitBurst: 0,
		RateLimit: 0,
	}
	conn := grpc.ClientConn{}
	invoker := func(currentCtx context.Context, currentMethod string, currentReq, currentRepl interface{}, currentConn *grpc.ClientConn, currentOpts ...grpc.CallOption) error {
		return nil
	}

	limiter := grpcclient.NewRateLimiter(&config)
	err := limiter(context.Background(), "methodName", "", "expectedReply", &conn, invoker)
	
	if se, ok := err.(interface {
		GRPCStatus() *status.Status
	}); ok {
		assert2.Equal(t, se.GRPCStatus().Code(), codes.ResourceExhausted)
		assert2.Equal(t, se.GRPCStatus().Message(), "rate: Wait(n=1) exceeds limiter's burst 0")
	} else{
		assert2.Fail(t, "Could not convert error into expected Status type")
	}
}
```